pandas/tests/test_strings.py

```diff
@@ -34,11 +188,115 @@ def test_api(self):
+        inferred_dtype, values = all_skipna_inferred_dtypes
+
+        t = box(values, dtype=dtype)  # explicit dtype to avoid casting
+
+        # TODO: get rid of these xfails
+        if dtype == 'category' and inferred_dtype in ['period', 'interval']:
+            pytest.xfail(reason='Conversion to numpy array fails because '
+                         'the ._values-attribute is not a numpy array for '
+                         'PeriodArray/IntervalArray; see GH 23553')
+        if box == Index and inferred_dtype in ['empty', 'bytes']:
+            pytest.xfail(reason='Raising too restrictively; '
+                         'solved by GH 23167')
+        if (box == Index and dtype == object
+                and inferred_dtype in ['boolean', 'date', 'time']):
+            pytest.xfail(reason='Inferring incorrectly because of NaNs; '
+                         'solved by GH 23167')
+        if (box == Series
+                and (dtype == object and inferred_dtype not in [
+                    'string', 'unicode', 'empty',
+                    'bytes', 'mixed', 'mixed-integer'])
+                or (dtype == 'category'
+                    and inferred_dtype in ['decimal', 'boolean', 'time'])):
+            pytest.xfail(reason='Not raising correctly; solved by GH 23167')
+
+        types_passing_constructor = ['string', 'unicode', 'empty',
+                                     'bytes', 'mixed', 'mixed-integer']
+        if inferred_dtype in types_passing_constructor:
+            # GH 6106
+            assert isinstance(t.str, strings.StringMethods)
+        else:
+            # GH 9184, GH 23011, GH 23163
+            with pytest.raises(AttributeError, match='Can only use .str '
+                               'accessor with string values.*'):
+                t.str
+            assert not hasattr(t, 'str')
+
+    @pytest.mark.xfail(reason='not correctly raising on master; '
+                       'solved by GH 23167')
+    def test_api_mi_raises(self):
+        mi = MultiIndex.from_arrays([['a', 'b', 'c']])
+        with pytest.raises(AttributeError, match='Can only use .str accessor '
+                           'with Index, not MultiIndex'):
+            mi.str
+        assert not hasattr(mi, 'str')
+
+    @pytest.mark.parametrize('dtype', [object, 'category'])
+    @pytest.mark.parametrize('box', [Series, Index])
+    def test_api_per_method(self, box, dtype,
+                            all_allowed_skipna_inferred_dtypes,
+                            all_string_methods):
+        # this test does not check correctness of the different methods,
+        # just that the methods work on the specified (inferred) dtypes,
+        # and raise on all others
+
+        # one instance of each parametrized fixture
+        inferred_dtype, values = all_allowed_skipna_inferred_dtypes
+        method_name, minimal_args = all_string_methods
+
+        # TODO: get rid of these xfails
+        if (method_name not in ['encode', 'decode', 'len']
```

Comment:
same comment as above

Reply:
These xfails will be gone, and then the test reads very clearly, IMO

Reply:
this test cannot be broken up as easily, because the allowed types depend on the method being checked!